### PR TITLE
microdude: init at 2.3

### PR DESCRIPTION
**Pull request by camelpunch**
###### Description of changes

Add the microdude application: a GTK Python program that interfaces with the Arturia Microbrute analogue synth.

I'm new to nix and nixos, and have [another open PR for a different audio program (shuriken)](https://github.com/NixOS/nixpkgs/pull/216837).

This one works fine and connects to my Microbrute when run with the following:

```
cd pkgs/applications/audio/microdude/
nix-build -E 'let pkgs = import <nixpkgs> { }; in pkgs.python3Packages.callPackage ./default.nix {}'
result/bin/microdude
```

But fails when run from the root of nixpkgs with the following:

```
$ nix-build -A microdude
# lots of happy-looking build output
$ result/bin/microdude 
Traceback (most recent call last):
  File "/nix/store/w03lfbrl2jasp75x16cqbbs1iz7j5nvy-microdude-2.3/bin/.microdude-wrapped", line 18, in <module>
    editor.main()
  File "/nix/store/w03lfbrl2jasp75x16cqbbs1iz7j5nvy-microdude-2.3/lib/python3.10/site-packages/microdude/editor.py", line 423, in main
    self.init_ui()
  File "/nix/store/w03lfbrl2jasp75x16cqbbs1iz7j5nvy-microdude-2.3/lib/python3.10/site-packages/microdude/editor.py", line 121, in init_ui
    self.main_window.connect(
TypeError: <Gtk.Window object at 0x7f7b882aef00 (GtkWindow at 0x1f66530)>: unknown signal name: delete-event
```

Have I done something weird with the GTK dependencies?

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by camelpunch** [comment]
I added `--ignore-theme-index` to the `gtk-update-icon-cache` invocation in the program's Makefile to avoid running into:
```
gtk-update-icon-cache /nix/store/5cqga13h540lc0n9vykwqpg2namrp3v5-microdude-2.3/share/icons
gtk-update-icon-cache: No theme index file.
```

File: pkgs/applications/audio/microdude/default.nix [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+, setproctitle
+, python-rtmidi
+, python3Packages
+}:
+
+python3Packages.buildPythonApplication rec {
+  pname = "microdude";
+  version = "2.3";
+
+  src = fetchgit {
```
Comment (AndersonTorres) [inline pkgs/applications/audio/microdude/default.nix:14]: Use fetchFromGitHub.

File: pkgs/applications/audio/microdude/default.nix [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    setproctitle
+    setuptools
+  ];
+
+  postInstall = ''
+    substituteInPlace Makefile \
+      --replace "python3 setup.py install" "" \
+      --replace gtk-update-icon-cache "gtk-update-icon-cache --ignore-theme-index" \
+      --replace '/usr/share/$$locale' $out/share/locale
+    make \
```
Comment (AndersonTorres) [inline pkgs/applications/audio/microdude/default.nix:40]: Shouldn't this be done at buildPhase?

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/applications/audio/microdude/default.nix (hunk 2 above)
Comment (camelpunch) [reply pkgs/applications/audio/microdude/default.nix:41]: Yup!

**Comment by camelpunch** [comment]
After making the above changes, I now get:

```
$ nix-env -f . -iA microdude
# happy-looking build output
$ microdude
Traceback (most recent call last):
  File "/nix/store/pjcc8ibw1alwqj8640ji985wg6j0k2hr-microdude-2.3/bin/.microdude-wrapped", line 4, in <module>
    from microdude.editor import Editor
ModuleNotFoundError: No module named 'microdude'
```

I'm guessing this is now a lack of proper runtime environment setup?

**Comment by camelpunch** [comment]
Thanks so much for your help BTW!

**Comment by camelpunch** [comment]
After a bit more digging, it seems that, since I remove the call to `python3 setup.py install` (avoiding all sorts of distutils and setuptools chaos), I need to reinstate the behaviour of including the `resources/*` dir in the output, in order for the executable to be able to import its modules. How I do that is eluding me...

**Comment by dagargo** [comment]
MicroDude includes a Python 3 module, which is installed via `setuptools`, and a few desktop related resources such as the desktop file, the application icon or the translations. As I didn't find a way to distribute these resources with `setuptools`, I decided to install them via `Makefile` and to reduce the amount of installation steps I decided to include a call to `python3 setup.py install` in the `install` target.

Here is the file for reference.

https://github.com/dagargo/microdude/blob/cc58e8e53919e6852131e1ea7a885a339e7dfb83/Makefile

But I discourage the packaging of MicroDude. I still need to add a warning to the project against its future usage. The reason is that [Elektroid](https://github.com/dagargo/elektroid) covers all the MicroDude functionality and much more. Elektroid is purely managed with `autotools` so packaging it would be easier.

Elektroid is already on Debian and includes installation instructions for other distributions.

Sadly, there is not an official Elektroid release including the compatibility with MicroBrute yet. But the master branch includes all the code and should work for everyone.

If you're interested, I can make the 2.5 release during the following days. (Some translations are missing.)

**Comment by camelpunch** [comment]
@dagargo ah great, I hadn't seen Elektroid, but it looks awesome. I'll close this PR and perhaps look at nixifying Elektroid.

**Pull request closed without merging**